### ISSUE TITLE
New func_door and func_button properties

### ISSUE THREAD

**Issue by bmorel** [Unvanquished/Unvanquished#2454]
While fixing #843 and even before that, we noticed (see #2453) that `+activate` can be used to open damage doors. This could very easily be promoted as a feature, only the interface would need to be defined. Notably, about the USE_DISTANCE problem.

Other points are, it seems NetRadiant's documentation states that the "health" property is kind of boolean, while I implemented it exactly like it would work a human. From IRC discussion, 2 ideas emerged:

* those could only receive damages that have above a certain number. This could be done either as "count only if higher than N", or as if "reduce incoming damages by N"
* it would make sense to have a regeneration speed

Ping @cu-kai as you're the only active mapper around those days. What do you think about those, would those have an interest to you?
From a technical point of view, they all seem fairly easy to implement to me, and since the code is still fresh in my mind, this is likely the ideal moment for that.

**Comment by bmorel** [comment]
One notable problem with the `+activate` problem is that this will depend on the difference between `USE_RANGE` and size of the entity to be used reliably. This is the only thing that might be a bit annoying to implement (mostly because of the talking though) as this constant is also used by the armoury, reactor and drill for human resupplies.
My opinion on this problem is, `USE_RANGE` should be a build-time constant, and each usable entity should have a use_distance property (specified by the .bsp or by configuration files for buildings). This use_distance property should be added to the "average radius" of the entity (that is: `half_width + sqrt ( half_width^2+half_width^2+ ) / 2` ) and would not be able to be more than `USE_RANGE`.
This would, I think, give a rather uniform feeling about the building reacting when close enough to the *perceived* building (perceived != bbox, especially since many stuff are cylinders) while giving decent performances (a distance is easy to calculate, and `USE_RANGE` would make sure that no ridiculous value would break performances).

**Comment by cu-kai** [comment]
I don't remember how the `health` property worked in tremulous and whether you needed to damage by X health, or simply deal any damage. You might do to look at the code for that.

I think it would be interesting to see a `+use` property for func_door, func_button, etc and I think it should _probably_ be implemented as a spawnflag to be toggled on/off (if on, "health" would be ignored, too).

**Comment by bmorel** [comment]
> (if on, "health" would be ignored, too).

I see no reason for such a limitation. I think it should be possible (and easier!) to open a door either gently, or by sending it a fully charged lucifer blast. They do that all the time on television :)

**Comment by bmorel** [comment]
I think the 2 ideas I mentioned in OP can be implemented together (they're not exclusive). It would also be possible to add a 3rd one.
The way I would see it is to simply add 2 key-values to the list of stuff currently supported:

1. regenerationRate, with a value indicating how many time would be needed for the mover to come from 0% hp to 100% hp. Or just how many hp to regen per tick, I guess.
2. shrugDamages, which would simply reduce all incoming damages by an integer value. This could be used for example so that a mover can only be destroyed or activated by a tyrant or lucifer cannon.
3. armor, with a percentage value, would reduce incoming damages by a percent. This is a bit like 2., except that it would not prevent smaller weapons to be successful at opening or destroying stuff.

As for the `+activate` feature, I have opened a PR for that. Still needs a map to test, and to iron things. @cu-kai if you do (or extend) such a test map, could you try to make `+activate` entities with various volumes or  more or less "plugged" in the wall (or floor)? I suspect the placement of things might make the entity unusable if we just use a flag to turn things on/off (I think station50 have the issue actually)

**Comment by cu-kai** [comment]
Support for the required bitflag for `+activate` was added to NetRadiant here: https://github.com/Unvanquished/unvanquished-mapeditor-support/commit/9030c282847be94b731bf14d7f3ad7f16cddc21a

**Comment by bmorel** [comment]
I'm still wondering about the 3 other ideas though:

* armour
* regen
* shrug off

they could be easy to implement to me, really, and might be useful for mappers. But I need opinion of mappers on that, such as you @cu-kai or @Viech  or any other one.

**Comment by bmorel** [comment]
I have not dug in the GUI side of damage and +activate features yet, because not much time plus I don't know cgame part much. For damages, I think it should be reasonably easy to give the player the same feedback as when hitting an alive entity. For +activate, I would just give the same feedback as when close to an armoury, but I don't really feel like investing time in this considering the fact the feature is not used yet.

As for feedback about how many damages something received, I think it can be done by mappers already the same way it's done for human buildings: adding smoke or cracks on the item is probably something about shaders to do. Now that I think about it, does alien buildings have such feedback? I think I only noticed it for human buildings (the smoke)?

**Comment by Viech** [comment]
> Now that I think about it, does alien buildings have such feedback? I think I only noticed it for human buildings (the smoke)?

Right; but that's also not ideal. Note that I'm not against the proposed features, I just don't think that I would use them in a map personally.

**Comment by bmorel** [comment]
Sure, no problem. That's kind of why I'm asking to actual mappers if they'd use those features instead of loosing time implementing them for nothing.
Myself not being a mapper (yet, but it's planned when trenchbroom will be usable for q3) makes me a bit unable to see what mappers would really need for map-entities features.

So far, the damage thing was actually a repair, and the +activate feature was extracted from a bug (it was possible to open damage triggered doors by using +activate) that's why I implemented those in sgame.

They certainly need more integration in cgame as well, but then cgame is not the prettiest part of the code, and I'm always a bit afraid to dig in that. It's also full of vec3_t and the like, that have yet to be migrated, and this does not helps my motivation, I have to admit.
